python/mujinwebstackclient/webstackclientutil.py

```diff
@@ -0,0 +1,295 @@
+        """Initialize all internal variables
+        """
+        self._queryFunction = queryFunction
+        self._args = args
+        self._kwargs = copy.deepcopy(kwargs)
+        self._items = []
+        self._shouldStop = False
+        self._kwargs.setdefault('offset', 0)
+        self._kwargs.setdefault('limit', 0)
+        self._totalLimit = self._kwargs['limit']
+        self._count = 0
+        if self._kwargs['limit'] > 0:
+            self._kwargs['limit'] = min(self._kwargs['limit'], maxQueryLimit)
+        else:
+            self._kwargs['limit'] = maxQueryLimit
+
+    def __iter__(self):
+        return self
+    
+    def __next__(self):
+        """Retrieve the next item from iterator
+           Required by Python3
+        """
+        return self.next()
+
+    def next(self):
+        """Retrieve the next item from iterator
+           Required by Python2
+        """
+
+        # return an item from internal buffer if buffer is not empty
+        if len(self._items) != 0:
+            item = self._items[0]
+            self._items = self._items[1:]
+            self._count += 1
+            return item
+
+        # stop iteration if internal buffer is empty and no need to query webstack again
+        if self._shouldStop:
+            raise StopIteration
+
+        # query webstack if buffer is empty
+        self._items = self._queryFunction(*self._args, **self._kwargs)
+        self._totalCount = self._items.totalCount
+        self._kwargs['offset'] += len(self._items)
+
+        if len(self._items) < self._kwargs['limit']:
+            # webstack does not have more items
+            self._shouldStop = True
+        if self._totalLimit != 0 and self._count + len(self._items) >= self._totalLimit:
+            # all remaining items user requests are in internal buffer, no need to query webstack again
+            self._shouldStop = True
+            self._items = self._items[:self._totalLimit - self._count]
+
+        return self.next()
+
+class QueryResult(list):
+    """Wraps query response. Break large query into small queries automatically to save memory.
+    """
+    _queryFunction = None # the actual webstack client query function (e.g. client.GetScenes)
+    _args = None # positional arguments supplied to the query function (e.g. scenepk)
+    _kwargs = None # keyword arguments supplied to the query function (e.g. offset=10, limit=20)
+    _meta = None  # meta dict returned from server
+    _items = None # internal buffer for items retrieved from webstack
+    _limit = None # query limit specified by the user
+    _offset = None # query offset specified by the user
+    _currentOffset = None # the offset for the first value inside buffer
+    _hasCompleteQueryResult = None # whether already has a complete list of query result
+
+    def __init__(self, queryFunction, *args, **kwargs):
+        self._queryFunction = queryFunction
+        self._args = args
+        self._kwargs = kwargs
```

Comment:
You should copy the kwargs during initializing instead of copying them at every iteration in `_APICall`